Add command for Typescript typedefs generation

Edo-San opened the pull request:
## Pull request type
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Other (please describe):

## How to test this PR
I have set up an npm `dev` script that can be useful to conveniently run the build step and the CLI in a "pseudo-dev mode".
- Log in
- Locate a file with a components JSON Schema (or generate one via the `pull-components` command)
- Run the `generate-typescript-typedefs` command providing the components JSON Schema via the `--sourceFilePaths` flag
- Inspect the generated Typescript definition file
- Try out different flags and check the output

### Command example
```
npm run dev generate-typescript-typedefs -- --sourceFilePaths ./components.12345.json
```

## What is the new behavior?
The main purpose of this PR is to add a command to the CLI that would provide our users with first-party support for generating Typescript type definitions based on Storyblok component schemas. As we're dealing with Typescript, it was a good opportunity to open up Typescript adoption in the CLI as well, as it made working with Typescript MUCH easier.

Here's a list of the provided changes:
- Create a `GenerateTypesFromJSONSchemas` class responsible for the whole rendering cycle of the Typescript type definitions file.
- Add support for Typescript files and build process with [`unbuild`](https://github.com/unjs/unbuild) - required updating all import types to be ESM
- Add `.vscode` settings to prevent automated formatting of existing `js` files while working on `ts` files
- Add a `generate-typescript-typedefs` CLI command and task file
- Add Typescript types for the whole task and its internals under the new directory `types`
- Add a section to the `README` file related to the `generate-typescript-typedefs` command

### TBD / Potential issues / Need hints
- As for now, the generated type definitions file is importing the `ISbStoryData` type from the same `storyblok` package, but there are some issues with exporting those types from this package, namely because some of them are also being imported from the `storyblok-js-client`. I'd assume that any user who is willing to use this new command is going to rely on the `storyblok-js-client`, at some point or another, so I guess the way to go is to somehow export the `types` alongside the `package.json` with `@storyblok/js` or `storyblok-js-client` listed as dependencies so that we can be sure we're going to import types that are going to be available from our distributed types. I still have to find out how to do that properly :D
- Maybe we need automated tests for this (read below)? There are a lot of options/combinations to handle, it's hard to test them all manually. Furthermore, updating CJS to ESM imports caused unit tests to broke, so I need to figure out how to fix them/refactor them.

### Potential improvements
As I had the chance to dig further into the CLI repo, I have outlined some potential improvements, for both the `generate-typescript-typedefs` command and the whole repo:
- [TS] Create a way to avoid clashes between `StoryblokProvidedPropertyTypes` and bloks (components) that can be created in Storyblok with the same name. I.e. a user can create a blok named `table`, and the types will end up having two `TableStoryblok` types, one for the user's component and one for the Storyblok-provided `Table` type.
- [TS] Give the possibility to the user to code the custom type parser in Typescript
- [TS] Add automated tests (need setup for TS tests) - Do you think this is required? It could also be a good opportunity to set them up with TS and to provide some mocks that could also be useful as internal examples, such as a `customFieldTypesParser` or a `JSONSchemaToTSOptions`.
- [CLI] Code formatting is missing
- [CLI] Unified error handling (maybe a dedicated class?)
- [CLI] Typescript support on all the codebase
- [CLI] There are undocumented commands, like `scaffold`, that are not documented in the README file. Maybe we could benefit from some cleanup. That command in particular I think is related to the old Rendering Service.

## Other information
A new `build` step should be added to trigger the `build` script.

This work has been heavily inspired by dohomi's [Storyblok generate ts](https://github.com/dohomi/storyblok-generate-ts). Thank you for your hard work and support! 🙏

Edo-San committed:
feat: add json-schema-to-typescript dep

Edo-San committed:
feat: add vite and boilerplate ts script

Edo-San committed:
feat: add ts-node for developing and unbuild for building process

Edo-San committed:
feat: add bridge file between cli parser and actual converting function

Edo-San committed:
chore: remove package-lock

Edo-San committed:
feat: update build to ESM and update all imports/exports. Disable typedefs generation

Edo-San committed:
feat: update typings and refactor parseBlokSchemaObject function

Edo-San committed:
refactor: refactor schemaobject type switch

Edo-San committed:
fix: update typeMapper func to correctly generate bloks allowlists

Edo-San committed:
fix: fix restrictions on bloks array fields for explicit list of bloks and no restrictions cases

Edo-San committed:
fix: update compile bannercomments and fix usage of custom type parser

Edo-San committed:
refactor: move types to dedicated folder

Edo-San committed:
refactor: refactor types, remove unused cli type

Edo-San committed:
refactor: update CLI params, refactor types

Edo-San committed:
feat: refactor types and add class baseline

Edo-San committed:
feat: add all methods to TS generation class

Edo-San committed:
refactor: get rid of old pieces of code and update calling new class methods

Edo-San committed:
chore: set all props and methods of the generator class as private

Edo-San committed:
chore: add comments to class methods

Edo-San committed:
chore: improve vars and functions naming consistency

Edo-San committed:
chore: move all types to typedef file

Edo-San committed:
feat: add readme basic docs, fix custom type parser path param

Edo-San committed:
feat: update readme

Edo-San committed:
feat: add dev npm script

Edo-San committed:
chore: get rid of unneeded files

Edo-San committed:
chore: update generator class comment

Edo-San committed:
feat: add type module to package json

Edo-San committed:
fix: update tests so that they could run with ESM

Edo-San committed:
fix: prevent tests from writing migration files that caused test errors

Edo-San commented:
#### Update 23 Feb 2024
I refactored all the tests so that they could run with ESM.

Edo-San committed:
feat: add declaration files

Edo-San committed:
feat: bump js client dep

Edo-San commented:
#### Update 28 Feb 2024

I bumped the `storyblok-js-client` version to `6.7.1` because that includes exporting a TS interface that was hindering the build process ATM. Now running `npm run build` should successfully build the CLI Node executable and it should also export its type definitions.

ademarCardoso commented on tests/units/sync-components.spec.js:
```diff
@@ -170,49 +172,40 @@ const mockPostRequest = jest.fn((path, payload) => {
+//       return originalModule.filterPresetsFromTargetComponent(presets, targetPresets)
+//     })
+//   }
+// })
```
I think you could delete this

ademarCardoso commented on src/tasks/generate-typescript-typedefs.ts:
```diff
@@ -0,0 +1,79 @@
+
+const generateTypescriptTypedefs: GenerateTSTypedefs = async ({
+  sourceFilePaths,
+  destinationFilePath = "./storyblok-component-types.d.ts",
```
I recommend that to generate the new file, you add the space number, as is done in the pull-components command.
```sh
# To generate files like that...
./storyblok-component-types_SPACE_ID.d.ts
```

I recommend this because many users have components for several spaces and without it the script will overwrite itself every time the user runs the command.

Edo-San replied:
I can see your point, but I am not entirely sure about that 🤔 
There isn't a direct correlation between Typescript type definitions and spaces, unlike the relationship between components JSONSchemas and spaces.

Those type definitions can host multiple types coming from different sources. The `sourceFilePaths` is indeed plural - notice the "s" - so that a user can pull multiple sources of components (that could be both single components or a whole group of components coming from a single space) and have them all included in the same definitions. For this reason, I think that labeling the typedefs file with the space ID by default could be a bit deceiving.

A couple of positive side effects are also deduplication (you won't have multiple `AssetStoryblok` provided types) and ease of import (one single source of import for all the Storyblok component types).

WDYT?

ademarCardoso replied:
I see, it makes sense, so I think we can keep it and in the future we can improve it if necessary.

ademarCardoso reviewed (commented):
This is HUGE, amazing refactor @Edo-San 🚀 

Please take a [look in this comment](https://github.com/storyblok/storyblok-cli/pull/79/files#r1513562539), it's very important.

Just to make a note:

**This PR will generate a breaking change, it will be necessary to publish as V4.0.0**

Edo-San committed:
chore: remove comment from spec file

Edo-San committed:
fix: solve conflicts on cli.ts

Edo-San committed:
fix: fix new tests

Edo-San commented:
> This is HUGE, amazing refactor @Edo-San 🚀
> 
> Please take a [look in this comment](https://github.com/storyblok/storyblok-cli/pull/79/files#r1513562539), it's very important.
> 
> Just to make a note:
> 
> **This PR will generate a breaking change, it will be necessary to publish as V4.0.0**

Thank you @ademarCardoso Appreciate! 🤗 

I wanted to ask you about the breaking changes, as I am probably missing something. Aside from the fact that with this new build step, there won't be the possibility to directly import anything from the CLI, as we won't be exposing all the CLI files anymore, do you see any other breaking changes?

#### UPDATE 06/03/2024
I rebased the PR on the updated master branch and fixed the conflicts related to the merged #81 PR

ademarCardoso commented:
> > This is HUGE, amazing refactor @Edo-San 🚀
> > Please take a [look in this comment](https://github.com/storyblok/storyblok-cli/pull/79/files#r1513562539), it's very important.
> > Just to make a note:
> > **This PR will generate a breaking change, it will be necessary to publish as V4.0.0**
> 
> Thank you @ademarCardoso Appreciate! 🤗
> 
> I wanted to ask you about the breaking changes, as I am probably missing something. Aside from the fact that with this new build step, there won't be the possibility to directly import anything from the CLI, as we won't be exposing all the CLI files anymore, do you see any other breaking changes?
> 
> #### UPDATE 06/03/2024
> I rebased the PR on the updated master branch and fixed the conflicts related to the merged #81 PR

Due to the amount of changes in the project, I thought it would be a breaking change, but I talked to other engineers and since the cli itself won't change the user's behavior, in other words, I don't think anything will break, so it's not a breaking change.

ademarCardoso commented:
@Edo-San I don't know if there is a QA on your team, but I would recommend that this PR be reviewed by one, to ensure that everything is working as it should. Carlos is an excellent QA and has experience with the CLI

carlosericciardi commented:
There is an error when the component has numbers at the beginning of its name:
![image](https://github.com/storyblok/storyblok-cli/assets/96299684/48db6fcc-8014-4551-9ee9-9184b5fc5627)

command: npm run dev generate-typescript-typedefs -- --sourceFilePaths ./components.139799.json

Edo-San committed:
feat: add logic to prepend an underscore if a type name starts with a number, update gitignore

Edo-San committed:
chore: add todo comment to keep in mind edge cases

Edo-San commented:
> There is an error when the component has numbers at the beginning of its name: ![image](https://private-user-images.githubusercontent.com/96299684/311726932-48db6fcc-8014-4551-9ee9-9184b5fc5627.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAxNzI3NTIsIm5iZiI6MTcxMDE3MjQ1MiwicGF0aCI6Ii85NjI5OTY4NC8zMTE3MjY5MzItNDhkYjZmY2MtODAxNC00NTUxLTllZTktOTE4NGI1ZmM1NjI3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzExVDE1NTQxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg3YTIzZGE5OTA5YzNmNGY0MGI1OTU3ODUzYmE2MTY2NjM4NWYzY2ZjNDNkOTgyMjQzYTNjZWE5OWE5ZjNhOWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.KJ4VKR_vJSJa6ME7lX6QvxKJjIFXs_4MXR2wSK6e3wk)
> 
> command: npm run dev generate-typescript-typedefs -- --sourceFilePaths ./components.139799.json

Pushed a fix for this one!
Basically, Typescript identifiers cannot start with a digit. I added some logic that, after all the transformations, check if the type name starts with a number and, if that's the case, it appends an underscore to the type name.

Thanks again for pointing that out @carlosericciardi 🙏

carlosericciardi reviewed (approved):
Awesome work!

Edo-San commented:
@ademarCardoso Do you think we can merge it now that @carlosericciardi has reviewed it? You can merge it straight away if you want 🤗

Edo-San commented:
Merging it as we good greenlight from @gustavomelki 

### UPDATE
Reverted because a build step is missing on the CI
